flair/nn/model.py

```diff
@@ -628,7 +673,24 @@ def _calculate_loss(self, scores, labels) -> Tuple[torch.Tensor, int]:
                 device=flair.device,
             )
 
-        return self.loss_function(scores, labels), len(labels)
+    def forward_loss(self, sentences: List[DT]) -> Tuple[torch.Tensor, int]:
+
+        # make a forward pass to produce embedded data points and labels
+        predict_data_points = self._get_prediction_data_points(sentences)
+        labels = self._prepare_label_tensor(predict_data_points)
+
+        # no loss can be calculated if there are no labels
+        if labels.size(0) == 0:
+            return torch.tensor(0.0, requires_grad=True, device=flair.device), 1
+
+        embedded_tensor = self._prepare_tensors(sentences)
```

Comment:
This could cause some problems I think. 4 lines above the `predict_data_points` are extracted using `self._get_prediction_data_points(sentences)` and labels extracted from them. 

But here now the `sentences` are passed into `self._prepare_tensors` which first applies a filter, then again calls `self._get_prediction_data_points(sentences)`. So it is possible that the datapoints extracted after the filter diverge from the datapoints used to get the labels. 

Maybe this `_prepare_tensors` could take as input `predict_data_points` instead of `sentences`? The filtering could be done beforehand as first line in `forward_loss`.

Reply:
Very good catch!
Although the filtering is usually for sentences that do not have any labels, it could be misused and create some divergence.

I moved the filtering to be the first thing to be done to a batch. Even before calling `_prepare_tensors`.

Sadly, I cannot change the signature of `_prepare_tensors`, as otherwise it wouldn't be in line with the general `flair.nn.Model` and would make it very complicated to use jit or onnx exports.